src/assoc_array.d

```diff
@@ -0,0 +1,1046 @@
+    // xtoHash, xopEquals, xopCmp, xtoString and xpostblit
+    ti.m_RTInfo = null;
+    immutable entrySize = talign(kti.tsize, vti.talign) + vti.tsize;
+    ti.m_init = (cast(ubyte*) null)[0 .. entrySize]; // init length, but not ptr
+
+    // xdtor needs to be built from the dtors of key and value for the GC
+    ti.xdtorti = &entryDtor;
+
+    ti.m_flags = TypeInfo_Struct.StructFlags.isDynamicType;
+    ti.m_flags |= (keyti.flags | valti.flags) & TypeInfo_Struct.StructFlags.hasPointers;
+    ti.m_align = cast(uint) max(kti.talign, vti.talign);
+
+    return ti;
+}
+
+//==============================================================================
+// Helper functions
+//------------------------------------------------------------------------------
+
+private size_t talign(size_t tsize, size_t algn) @safe pure nothrow @nogc
+{
+    immutable mask = algn - 1;
+    assert(!(mask & algn));
+    return (tsize + mask) & ~mask;
+}
+
+// mix hash to "fix" bad hash functions
+private size_t mix(size_t h) @safe pure nothrow @nogc
+{
+    // final mix function of MurmurHash2
+    enum m = 0x5bd1e995;
+    h ^= h >> 13;
+    h *= m;
+    h ^= h >> 15;
+    return h;
+}
+
+private size_t calcHash(in void* pkey, in TypeInfo keyti)
+{
+    immutable hash = keyti.getHash(pkey);
+    // highest bit is set to distinguish empty/deleted from filled buckets
+    return mix(hash) | HASH_FILLED_MARK;
+}
+
+private size_t nextpow2(in size_t n) pure nothrow @nogc
+{
+    import core.bitop : bsr;
+
+    if (!n)
+        return 1;
+
+    const isPowerOf2 = !((n - 1) & n);
+    return 1 << (bsr(n) + !isPowerOf2);
+}
+
+pure nothrow @nogc unittest
+{
+    //                            0, 1, 2, 3, 4, 5, 6, 7, 8,  9
+    foreach (const n, const pow2; [1, 1, 2, 4, 4, 8, 8, 8, 8, 16])
+        assert(nextpow2(n) == pow2);
+}
+
+private T min(T)(T a, T b) pure nothrow @nogc
+{
+    return a < b ? a : b;
+}
+
+private T max(T)(T a, T b) pure nothrow @nogc
+{
+    return b < a ? a : b;
+}
+
+//==============================================================================
+// API AssocArrayementation
+//------------------------------------------------------------------------------
+
+/// Determine number of entries in associative array.
+
+extern (C) size_t _aaLen(in AssocArray!(void, void)* aa) pure nothrow @nogc
+{
+    return aa ? aa.length : 0;
+}
+
+/******************************
+ * Lookup *pkey in aa.
+ * Called only from implementation of (aa[key]) expressions when value is mutable.
+ * Params:
+ *      aa = associative array opaque pointer
+ *      ti = TypeInfo for the associative array
+ *      valsz = ignored
+ *      pkey = pointer to the key value
+ * Returns:
+ *      if key was in the aa, a mutable pointer to the existing value.
+ *      If key was not in the aa, a mutable pointer to newly inserted value which
+ *      is set to all zeros
+ */
+extern (C) void* _aaGetY(AssocArray!(void, void)** aa, const TypeInfo_AssociativeArray ti, in size_t valsz,
```

Comment:
maybe use `ref AssocArray!(void, void)* aa` here to avoid repeating `(*aa)` below. Or assign it to some local.